docs/src/tutorial/pxd_files.rst

```diff
@@ -1,48 +1,81 @@
 .. _pxd_files:
 
+*********
 pxd files
-=========
+*********
 
-In addition to the ``.pyx`` source files, Cython uses ``.pxd`` files
+.. include::
+    ../two-syntax-variants-used
+
+In addition to the ``.pyx`` and ``.py`` source files, Cython uses ``.pxd`` files
 which work like C header files -- they contain Cython declarations
 (and sometimes code sections) which are only meant for inclusion by
-Cython modules.  A ``pxd`` file is imported into a ``pyx`` module by
+Cython modules.  A ``.pxd`` file is imported into a ``.pyx`` module by
 using the ``cimport`` keyword.
 
-``pxd`` files have many use-cases:
+``.pxd`` files have many use-cases:
 
 1.  They can be used for sharing external C declarations.
 2.  They can contain functions which are well suited for inlining by
-    the C compiler. Such functions should be marked ``inline``, example::
+    the C compiler. Such functions should be marked ``inline``, example:
+
+    .. tabs::
+
+        .. group-tab:: Pure Python
+
+            .. literalinclude:: ../../examples/tutorial/pxd_files/inline.py
```

Comment:
This seems borderline misleading. The inline function should really be in a `.pxd` file. It's in a `.pyx` file instead to make it testable. But having an additional `.py` file? I'm not even sure if it works to use `@cython.inline` with a `def` function in a `.pxd` file. And it's definitely not tested by this setup.

How would a reasonable setup look here? Should users have their code duplicated in a `.py` file (to support Python execution and code checkers) and in a `.pxd` file (to allow inlining it in Cython)?

Reply:
> This seems borderline misleading. The inline function should really be in a `.pxd` file.

Oh, I'm sorry, I, apparently, did not read the text well enough. 😅

> I'm not even sure if it works to use `@cython.inline` with a `def` function in a `.pxd` file.

Hmm, I will need to test everything with pxd.

> It's in a `.pyx` file instead to make it testable.

If I put them in pxd, will the compilation be checked anyway? As far as I know, the files in the documentation won't run anyway, so we can probably change the file extension to pxd.

> and in a `.pxd` file (to allow inlining it in Cython)?

Cython doesn't support inlining for py?

> Should users have their code duplicated in a `.py` file (to support Python execution and code checkers)

Well, duplication is definitely not welcome.

> How would a reasonable setup look here?

What? I don't really understand your question.

Reply:
>> and in a .pxd file (to allow inlining it in Cython)?
>
> Cython doesn't support inlining for py?

It doesn't support `def` functions in `.pxd` files.

>> How would a reasonable setup look here?

I was asking how users should deal with a setting in which they write everything in `.py` files and develop, test, run everything in Python, but want to provide inline functions, which are only available in `.pxd` files which Python tools do not see and understand. Should they duplicate their code for that? In the end, inlining is just an optimisation.

Reply:
Can we just keep the inline function in a `.pxd` file and not have two variants for it?

Reply:
> Can we just keep the inline function in a `.pxd` file and not have two variants for it?

Absolutely. This pr will be much more useful if it were merged even if it didn't cover every bit of code

> How would a reasonable setup look here?

I probably didn't understood the problem you were trying to discuss. My understanding is that if we declare cfunc in .py it should be identical to declaring it in the pxd and cimporting it.

Reply:
> My understanding is that if we declare cfunc in .py it should be identical to declaring it in the pxd and cimporting it.

Ah. Hmm. But then it's not inlined across modules. You need the `.pxd` file in order to make the implementation available to other modules that cimport it. And if you have the implementation both in the `.py` file (local to the module) and its `.pxd` file (for cross-module inlining), then its duplicated code. I think we should just clearly state that if users want to use an inline function across modules, then it needs to be in a `.pxd` file. (That, minus the "cimport-from-pyx" feature, but I don't know if that actually works for inline functions…)

Reply:
So, the problem is that users don't distribute the py, but they distribute the pxd?

Reply:
Yes, they can do that. In fact, there are packages on PyPI that only consist of `.pxd` files with declarations. And those might happily contain inline functions that then become part of the modules that cimport them.

(As a side note: It might actually be possible to add `.py` files to such packages with Python stubs in them, which could then be used by Python code running uncompiled in Python. But again, those wouldn't be inline functions, just plain Python functions. They wouldn't be used by compiled code, nor when it is being compiled by Cython.)

Reply:
resolved 40adc4c510932fea307d38e341ce382a6af233c1